### PR TITLE
Allows `KerasMetricCallback` to use XLA generation

### PR DESCRIPTION
**Pull request by Rocketknight1**
Updates the `KerasMetricCallback` with the ability to use XLA generation for a big speed boost! cc @merveenoyan

### CODE REVIEW

File: src/transformers/keras_callbacks.py [hunk 1]
```diff
@@ -65,6 +65,15 @@ def rouge_fn(predictions, labels):
             Batch size. Only used when the data is not a pre-batched `tf.data.Dataset`.
         predict_with_generate (`bool`, *optional*, defaults to `False`):
             Whether we should use `model.generate()` to get outputs for the model.
+        use_xla_generation (`bool`, *optional*, defaults to `False`):
+            If we're generating, whether to compile model generation with XLA. This can massively increase the speed of
+            generation (up to 100X speedup) but will require a new XLA compilation for each input shape. When using XLA
+            generation, it's a good idea to pad your inputs to the same size, or to use the `pad_to_multiple_of`
+            argument in `to_tf_dataset()`, which will reduce the number of unique input shapes and save a lot of
+            compilation time. This option has no effect is `predict_with_generate` is `False`.
```
Comment (gante) [inline src/transformers/keras_callbacks.py:73]: This docstring is great!

**Review by gante — APPROVED** [review]
👍 🔥

File: src/transformers/keras_callbacks.py [hunk 2]
```diff
@@ -126,6 +137,11 @@ def __init__(
         if parse(tf.__version__) < parse("2.7"):
             logging.warning("TF versions less than 2.7 may encounter issues with KerasMetricCallback!")
 
+        self.use_xla_generation = use_xla_generation
+        self.generate_kwargs = generate_kwargs or {}
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:141]: We don't use `or` in Transformers like this. Please be explicit with a one-line test.

File: src/transformers/keras_callbacks.py [hunk 3]
```diff
@@ -183,6 +199,17 @@ def on_epoch_end(self, epoch, logs=None):
...
                 main_input_name = getattr(self.model, "main_input_name", "input_ids")
 
+            if self.use_xla_generation and self.generation_function is None:
+
+                def generation_function(inputs, attention_mask):
+                    return self.model.generate(
+                        inputs,
+                        attention_mask=attention_mask,
+                        **self.generate_kwargs,
+                    )
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:209]: Fits in one line.

File: src/transformers/keras_callbacks.py [hunk 4]
```diff
@@ -76,6 +85,8 @@ def __init__(
         label_cols: Optional[List[str]] = None,
         batch_size: Optional[int] = None,
         predict_with_generate: Optional[bool] = False,
+        use_xla_generation: Optional[bool] = False,
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:88]: This argument is not Optional has it can't be None (same for the one above).

**Review by sgugger — COMMENTED** [review]
Had some nits :-)

**Pull request merged**